### PR TITLE
Allow OmniSharp to provide ImplementType Options

### PR DESCRIPTION
**Pull request by JoeRobich**
This should allow OmniSharp to provide ImplementType Options at the workspace level. Something similar to their [CSharpFormattingWorkspaceOptionsProvider](https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Roslyn.CSharp/Services/CSharpFormattingWorkspaceOptionsProvider.cs#L14).  Will be inserted into O# as part of https://github.com/OmniSharp/omnisharp-roslyn/pull/2642

### CODE REVIEW

**Comment by JoeRobich** [comment]
@tmat Could you take a look? You know this area better than most.

File: src/Features/ExternalAccess/OmniSharp/Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.csproj [hunk 1]
```diff
@@ -10,13 +10,14 @@
...
   </PropertyGroup>
-  
+
   <ItemGroup>
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.CSharp" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.UnitTests" />
     <!--
       ⚠ ONLY OMNISHARP ASSEMBLIES MAY BE ADDED HERE ⚠
     -->
+    <InternalsVisibleTo Include="OmniSharp.Host" Key="$(OmniSharpKey)" />
```
Comment (JoeRobich) [inline src/Features/ExternalAccess/OmniSharp/Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.csproj:20]: O# keeps the workspace options updated in this assembly.

File: src/Features/ExternalAccess/OmniSharp/Options/OmniSharpEditorConfigOptions.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.ImplementType;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.Options;
+
+internal sealed record class OmniSharpEditorConfigOptions
```
Comment (JoeRobich) [inline src/Features/ExternalAccess/OmniSharp/Options/OmniSharpEditorConfigOptions.cs:9]: For the time being these are the only fallback options that have matching O# options

**Review by JoeRobich — COMMENTED** [review]
@tmat I updated based on your suggestions. Please review.

File: src/Features/ExternalAccess/OmniSharp/Options/OmniSharpEditorConfigOptions.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+{
+    public OmniSharpLineFormattingOptions LineFormattingOptions { get; init; } = new();
```
Comment (tmat) [inline src/Features/ExternalAccess/OmniSharp/Options/OmniSharpEditorConfigOptions.cs:11]: Do we need these to be initialized to `new()`?

File: src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        try
+        {
+            workspace.SetCurrentSolution(UpdateOptions, changeKind: WorkspaceChangeKind.SolutionChanged);
+
+            Solution UpdateOptions(Solution oldSolution)
+            {
+                var oldFallbackOptions = oldSolution.FallbackAnalyzerOptions;
+                oldFallbackOptions.TryGetValue(LanguageNames.CSharp, out var csharpFallbackOptions);
+
+                var changedOptions = DetermineChangedOptions(csharpFallbackOptions, editorConfigOptions);
```
Comment (tmat) [inline src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs:31]: Do we need to determine changed options? Why not set all the O# options unconditionally?

Comment (JoeRobich) [reply src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs]: No, that simplifies a lot. Thanks!

**Comment by JoeRobich** [comment]
@tmat updated! Thanks

File: src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+using Workspace = CodeAnalysis.Workspace;
+
+internal static class OmniSharpSolutionAnalyzerConfigOptionsUpdater
+{
+    internal static void UpdateOptions(Workspace workspace, OmniSharpEditorConfigOptions editorConfigOptions)
+    {
+        try
+        {
+            workspace.SetCurrentSolution(UpdateOptions, changeKind: WorkspaceChangeKind.SolutionChanged);
```
Comment (CyrusNajmabadi) [inline src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs:24]: oh.  this is... concerning.  @tmat  for thoughts on someone actively calling into SetCurrentSolution. 

Is tehre a reason they need this?  do they not have their own Workspace suclass they can call through to do this?

File: src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Options;
+using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.ImplementType;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.Options;
+
+using Workspace = CodeAnalysis.Workspace;
```
Comment (CyrusNajmabadi) [inline src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs:16]: why is this needed?

File: src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Options;
+using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.ImplementType;
```
Comment (CyrusNajmabadi) [inline src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs:12]: sort.

File: src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs (hunk 6 above)
Comment (JoeRobich) [reply src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs:16]: Workspace conflicts because there is a `Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.Workspace` namespace.

File: src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs (hunk 5 above)
Comment (JoeRobich) [reply src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs:24]: Looking at `Workspace.TryApplyChanges` and it appears to copy the `FallbackAnalyzerOptions`. Perhaps I can just switch to using that instead.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        static void AddOption<T>(
+            PerLanguageOption2<T> option,
+            T value,
+            ImmutableDictionary<string, string>.Builder builder)
+        {
+            var configName = option.Definition.ConfigName;
+            var configValue = option.Definition.Serializer.Serialize(value);
+            builder.Add(configName, configValue);
```
Comment (tmat) [inline src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs:71]: Should this not replace the value if it already exists?

File: src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        catch (Exception e) when (FatalError.ReportAndPropagate(e, ErrorSeverity.Diagnostic))
+        {
+            throw ExceptionUtilities.Unreachable();
+        }
+
+        static void AddOption<T>(
+            PerLanguageOption2<T> option,
+            T value,
+            ImmutableDictionary<string, string>.Builder builder)
```
Comment (tmat) [inline src/Features/ExternalAccess/OmniSharp/Options/OmniSharpSolutionAnalyzerConfigOptionsUpdater.cs:67]: nit: you can make this function non-static and remove builder param.
